### PR TITLE
closes #387 - cli for `assert_filter_cond()`

### LINKED ISSUE #387: `assert_filter_cond()`

### PR DESCRIPTION
**Pull request by ddsjoberg**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/dev/CONTRIBUTING.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #387 into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiraldev/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: R/assertions.R
Comment (bundfussr) [inline R/assertions.R:1]: Could we update the example to
```
#' example_fun <- function(dat, x) {
#'   x <- assert_filter_cond(enexpr(x), arg_name = "x")
#'   filter(dat, !!x)
#' }
```
The current example produces
```
#> Error in filter(dat, !!x) : ℹ In argument: `USUBJID`.
#> Caused by error:
#> ! `..1` must be a logical vector, not a character vector.
```
but not the error we want to show.

Comment (ddsjoberg) [reply R/assertions.R:1]: I think this is a bug in the function and unrelated to any messaging updates. Check out the example below that is using the versin of admiraldev from CRAN. `assert_filter_cond()` is _not_ throwing an error, rather the error is from `dplyr::filter()`.

``` r
library(admiraldev)
packageVersion("admiraldev")
#> [1] '1.0.0'

library(pharmaversesdtm)
library(dplyr, warn.conflicts = FALSE)
library(rlang)
#> 
#> Attaching package: 'rlang'
#> The following object is masked from 'package:admiraldev':
#> 
#>     is_named
data(dm)

# typical usage in a function as an argument check
example_fun <- function(dat, x) {
  x <- assert_filter_cond(enquo(x))
  filter(dat, !!x)
}

example_fun(dm, USUBJID)
#> Error in `filter()`:
#> ℹ In argument: `USUBJID`.
#> Caused by error:
#> ! `..1` must be a logical vector, not a character vector.

# assert is not the function returning the error
example_fun2 <- function(dat, x) {
  assert_filter_cond(enquo(x))
}
```

<sup>Created on 2024-03-14 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup> 


We probably need to update this function, and if we do, it seems that to properly test if an expression is a proper filter condition, we'd need to evaluate within a data mask to ensure that it evaluates to logical, e.g. `eval_tidy(expr, data = data) |> is.logical()`. (This would probably need to be wrapped in a tryCatch() as well.) The example above passing the column name `USUBJID` should fail. But a similar example where the column name passed is a column of `TRUE` and `FALSE` is proper notation.

What do you think the best way forward is here?

Comment (bundfussr) [reply R/assertions.R:1]: For now I would just update the example.

I think we should update the messages first and in the next step we can update the code where user specified expressions are evaluated. I would expect most of the updates in admiral.

I would not try to evaluate the filter condition in the assertions because
- the data might not be available when the assertion is called and
- we would evaluate the expression twice once in the assertion and again in the code of the calling function.

We could also think about if we need the assertion at all. It is quite weak and as you have pointed out it does not work for logical columns. So once we catch errors where the expression is evaluated we may not need to catch a few cases beforehand.

Comment (ddsjoberg) [reply R/assertions.R:1]: Ok thanks @bundfussr I updated the example, but I couldn't use the `example_fun()` defined in example section, because the input is a quosure which will always pass the assertion. I agree that re-visiting this function after these updates are done is a good idea. Doesn't seem to have the advertised functionality? I'll make a new issue.

``` r
rlang::is_call(rlang::quo(anything_can_go_here))
#> [1] TRUE
```

<sup>Created on 2024-03-15 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

File: R/assertions.R [hunk 1]
```diff
@@ -489,12 +495,12 @@ assert_filter_cond <- function(arg, optional = FALSE) {
...
   provided <- !is_missing(arg)
   if (provided && !(is_call(arg) || is_logical(arg))) {
-    err_msg <- sprintf(
-      "`%s` must be a filter condition but is %s",
-      arg_name(substitute(arg)),
-      what_is_it(arg)
+    cli_abort(
+      message = message %||%
+        "Argument {.arg {arg_name}} must be a filter condition, but is {.obj_type_friendly {arg}}",
+      class = c(class, "assert-admiraldev"),
```
Comment (bms63) [inline R/assertions.R:501]: The assert functions had their classes more specific to their assertions.  how come this one is `assert-admiraldev`.  Could it be `assert-filter_cond` to be consistent with the others?

**Review by bms63 — APPROVED** [review]
Is this PR in limbo because of the previous discussion?  It looks like those have been made into separate issues.  

I just had a question on the class for this function.

File: R/assertions.R (hunk 1 above)
Comment (bms63) [reply R/assertions.R:501]: Nevermind!! I see my misunderstanding!!
